### PR TITLE
[Kernel] Add w8a8 CUTLASS kernels

### PR DESCRIPTION
**Pull request by tlrmchlsmth**
This PR adds fp8_e4m3fn and int8 GEMM kernels, using NVIDIA CUTLASS and unit tests for them. The kernels are not used in this present PR, but are planned to be used in https://github.com/vllm-project/vllm/pull/4525.

The main contributions of this PR is the function `cutlass_scaled_mm_dq`:
* Supports symmetric quantized activations and weights
* The activations may be either per-tensor or per-token
* The weights may be either per-tensor or per output channel
* int8 is supported on Turing, Ampere, Lovelace, or Hopper
* fp8_e4m3 is supported on Ampere or Lovelace.
* Outputs can be either bfloat16 or fp16.

---

<details>

<summary><b> PR Checklist (Click to Expand) </b></summary>

<p>Thank you for your contribution to vLLM! Before submitting the pull request, please ensure the PR meets the following criteria. This helps vLLM maintain the code quality and improve the efficiency of the review process.</p>

<h3>PR Title and Classification</h3>
<p>Only specific types of PRs will be reviewed. The PR title is prefixed appropriately to indicate the type of change. Please use one of the following:</p>
<ul>
    <li><code>[Bugfix]</code> for bug fixes.</li>
    <li><code>[CI/Build]</code> for build or continuous integration improvements.</li>
    <li><code>[Doc]</code> for documentation fixes and improvements.</li>
    <li><code>[Model]</code> for adding a new model or improving an existing model. Model name should appear in the title.</li>
    <li><code>[Frontend]</code> For changes on the vLLM frontend (e.g., OpenAI API server, <code>LLM</code> class, etc.) </li>
    <li><code>[Kernel]</code> for changes affecting CUDA kernels or other compute kernels.</li>
    <li><code>[Core]</code> for changes in the core vLLM logic (e.g., <code>LLMEngine</code>, <code>AsyncLLMEngine</code>, <code>Scheduler</code>, etc.)</li>
    <li><code>[Hardware][Vendor]</code> for hardware-specific changes. Vendor name should appear in the prefix (e.g., <code>[Hardware][AMD]</code>).</li>
    <li><code>[Misc]</code> for PRs that do not fit the above categories. Please use this sparingly.</li>
</ul>
<p><strong>Note:</strong> If the PR spans more than one category, please include all relevant prefixes.</p>

<h3>Code Quality</h3>

<p>The PR need to meet the following code quality standards:</p>

<ul>
    <li>We adhere to <a href="https://google.github.io/styleguide/pyguide.html">Google Python style guide</a> and <a href="https://google.github.io/styleguide/cppguide.html">Google C++ style guide</a>.</li>
    <li>Pass all linter checks. Please use <a href="https://github.com/vllm-project/vllm/blob/main/format.sh"><code>format.sh</code></a> to format your code.</li>
    <li>The code need to be well-documented to ensure future contributors can easily understand the code.</li>
    <li>Include sufficient tests to ensure the project to stay correct and robust. This includes both unit tests and integration tests.</li>
    <li>Please add documentation to <code>docs/source/</code> if the PR modifies the user-facing behaviors of vLLM. It helps vLLM user understand and utilize the new features or changes.</li>
</ul>

<h3>Notes for Large Changes</h3>
<p>Please keep the changes as concise as possible. For major architectural changes (>500 LOC excluding kernel/data/config/test), we would expect a GitHub issue (RFC) discussing the technical design and justification. Otherwise, we will tag it with <code>rfc-required</code> and might not go through the PR.</p>

<h3>What to Expect for the Reviews</h3>

<p>The goal of the vLLM team is to be a <i>transparent reviewing machine</i>. We would like to make the review process transparent and efficient and make sure no contributor feel confused or frustrated. However, the vLLM team is small, so we need to prioritize some PRs over others. Here is what you can expect from the review process: </p>

<ul>
    <li> After the PR is submitted, the PR will be assigned to a reviewer. Every reviewer will pick up the PRs based on their expertise and availability.</li>
    <li> After the PR is assigned, the reviewer will provide status update every 2-3 days. If the PR is not reviewed within 7 days, please feel free to ping the reviewer or the vLLM team.</li>
    <li> After the review, the reviewer will put an <code> action-required</code> label on the PR if there are changes required. The contributor should address the comments and ping the reviewer to re-review the PR.</li>
    <li> Please respond to all comments within a reasonable time frame. If a comment isn't clear or you disagree with a suggestion, feel free to ask for clarification or discuss the suggestion.
 </li>
</ul>

<h3>Thank You</h3>

<p> Finally, thank you for taking the time to read these guidelines and for your interest in contributing to vLLM. Your contributions make vLLM a great tool for everyone! </p>

</details>

### CODE REVIEW

**Comment by jeejeelee** [comment]
Great work~

Has the vllm  community begun integrating Cutlass? Is this PR part of the official roadmap? 

Additionally, For the integration of Cutlass, is it based on the python module(https://github.com/vllm-project/vllm/pull/4525) or the method outlined in your PR?

**Comment by tlrmchlsmth** [comment]
Thanks @jeejeelee -- this PR is part of a larger project to add support for w8a8 quantization (which is on the Q2 roadmap  https://github.com/vllm-project/vllm/issues/3861). We ran into several issues with the Python interface in #4525, and it's really not supposed to be used this way, so we plan to replace the python cutlass code with these C++ kernels.

The main reason for using CUTLASS here is its ability to do operator fusion via its epilogue operations. For int8 quantization, especially in the asymmetric case, there are a variety of small operations that we'd like to fuse onto GEMMs to avoid the cost of sweeping over the outputs multiple times (see https://github.com/vllm-project/vllm/issues/3975). In the future,

**Comment by tlrmchlsmth** [comment]
@pcmoritz @comaniac There are a couple of issues to iron out still (CMakeLists changes and kernel dispatching for sure) but this should be ready to look at.

@youkaichao do you have any advice on how to handle the SM90a issues? (I know you were looking into this -- unfortunate that https://github.com/pytorch/pytorch/commit/6e99f739235980e8d47e8fe6246c7466f2ce2f58 didn't make it into 2.3)

File: csrc/quantization/cutlass/common.hpp [hunk 1]
```diff
@@ -0,0 +1,47 @@
...
+/**
+ * Panic wrapper for unwinding CUTLASS errors
+ */
+#define CUTLASS_CHECK(status)                                                                    \
+  {                                                                                              \
+    cutlass::Status error = status;                                                              \
+    if (error != cutlass::Status::kSuccess) {                                                    \
+      std::cerr << "Got cutlass error: " << cutlassGetStatusString(error) << " at: " << __LINE__ \
+                << std::endl;                                                                    \
+      exit(EXIT_FAILURE);                                                                        \
```
Comment (pcmoritz) [inline csrc/quantization/cutlass/common.hpp:45]: Can we change this to call TORCH_CHECK so it raises a Python exception and doesn't simply quit the process? That's what we do everywhere else too :)

File: CMakeLists.txt [hunk 2]
```diff
@@ -16,7 +16,8 @@ include(${CMAKE_CURRENT_LIST_DIR}/cmake/utils.cmake)
 set(PYTHON_SUPPORTED_VERSIONS "3.8" "3.9" "3.10" "3.11")
 
 # Supported NVIDIA architectures.
-set(CUDA_SUPPORTED_ARCHS "7.0;7.5;8.0;8.6;8.9;9.0")
+set(CUDA_SUPPORTED_ARCHS "7.0;7.5;8.0;8.6;8.9;9.0;9.0a")
+set(CMAKE_CUDA_FLAGS "-gencode arch=compute_90a,code=sm_90a ${CMAKE_CUDA_FLAGS}")
```
Comment (bnellnm) [inline CMakeLists.txt:20]: Why did you need to modify `CMAKE_CUDA_FLAGS`?  Does torch not detect 9.0a?

Comment (tlrmchlsmth) [reply CMakeLists.txt:20]: No it doesn't unfortunately -- see https://github.com/pytorch/pytorch/commit/6e99f739235980e8d47e8fe6246c7466f2ce2f58 

I don't think this is the right fix, though so I am open to suggestions.

**Comment by tlrmchlsmth** [comment]
> Thank you for your patient explanation. May I ask another question?
> 
> Why isn't SM75 supported? We should be able to utilize the [m8n8k16](https://github.com/NVIDIA/cutlass/blob/v3.5.0/include/cutlass/arch/mma_sm75.h#L270)

I'll grab a T4 and see if I can get it working there

File: csrc/quantization/cutlass/cutlass_visitor_2x_broadcast_epilogue.hpp [hunk 3]
```diff
@@ -0,0 +1,332 @@
...
+#include "cute/tensor.hpp"
+
+// clang-format on 
+
+namespace cutlass::epilogue::threadblock {
+
+using namespace cute;
+using namespace detail;
+
+using X = Underscore;
```
Comment (bnellnm) [inline csrc/quantization/cutlass/cutlass_visitor_2x_broadcast_epilogue.hpp:55]: Is this alias used anywhere?

File: csrc/quantization/cutlass/cutlass_visitor_2x_broadcast_epilogue.hpp [hunk 4]
```diff
@@ -0,0 +1,332 @@
...
+        VecType filled_vec;
+        CUTLASS_PRAGMA_UNROLL
+        for (int i = 0; i < VecLength; i++) {
+          reinterpret_cast<Element*>(&filled_vec)[i] = params_ptr->null_default;
+        }
+
+        CUTLASS_PRAGMA_UNROLL
+        for (int i = 0; i < size(src_v); ++i) {
+          dst_v(i) = filled_vec;
+        }
```
Comment (bnellnm) [inline csrc/quantization/cutlass/cutlass_visitor_2x_broadcast_epilogue.hpp:146]: Does this loop need some sort of guard like the vector broadcast case?

File: csrc/quantization/cutlass/scaled_mm_dq_entry.cu [hunk 5]
```diff
@@ -0,0 +1,42 @@
...
+    // TODO: This kernel only works for sm90a
+    //  -- figure out how to detect 90a vs 90
+
+    cutlass_scaled_mm_dq_sm90(out, a, b, a_scales, b_scales);
+  } else if (version_num == 89) /* Ada Lovelace */ {
+    cutlass_scaled_mm_dq_sm89(out, a, b, a_scales, b_scales);
+  } else if (version_num >= 80) /* Ampere */ {
+    cutlass_scaled_mm_dq_sm80(out, a, b, a_scales, b_scales);
+  } else {
+    throw std::runtime_error("Unsupported GPU architecture");
```
Comment (bnellnm) [inline csrc/quantization/cutlass/scaled_mm_dq_entry.cu:40]: Use TORCH_CHECK here too?

File: csrc/quantization/cutlass/scaled_mm_dq_sm8x.cu [hunk 6]
```diff
@@ -0,0 +1,217 @@
...
+#include "cutlass/epilogue/threadblock/fusion/visitors.hpp"
+#include "cutlass/gemm/kernel/default_gemm_universal_with_visitor.h"
+
+#include "cutlass_visitor_2x_broadcast_epilogue.hpp"
+#include "common.hpp"
+// clang-format on
+
+/////////////////////////////////////////
+
+template <typename Arch, typename ElementAB_, typename ElementD_>
```
Comment (bnellnm) [inline csrc/quantization/cutlass/scaled_mm_dq_sm8x.cu:30]: Can you add a comment that describes the operation in human readable notation?  I think even a cut+paste of the equivalent python code would be helpful.

File: csrc/quantization/cutlass/scaled_mm_dq_sm90.cu [hunk 7]
```diff
@@ -0,0 +1,193 @@
...
+// clang-format on
+
+using namespace cute;
+
+/////////////////////////////////////////
+
+template <typename ElementAB_, typename ElementD_, typename TileShape,
+          typename ClusterShape, typename KernelSchedule,
+          typename EpilogueSchedule>
+struct sm90_gemm {
```
Comment (bnellnm) [inline csrc/quantization/cutlass/scaled_mm_dq_sm90.cu:30]: Ditto comment(s) from `scaled_mm_dq_sm8x.cu`

File: csrc/quantization/cutlass/scaled_mm_dq_sm90.cu [hunk 8]
```diff
@@ -0,0 +1,193 @@
...
+      cute::Shape<int, int, int, int>, CollectiveMainloop, CollectiveEpilogue,
+      cutlass::gemm::PersistentScheduler>;
+
+  struct GemmKernel : public KernelType {};
+};
+
+/////////////////////////////////////////
+
+using StrideA = Stride<int64_t, Int<1>, Int<0>>;
+using StrideB = Stride<int64_t, Int<1>, Int<0>>;
```
Comment (bnellnm) [inline csrc/quantization/cutlass/scaled_mm_dq_sm90.cu:103]: Seems like these aliases could be local to the dispatcher?

File: csrc/quantization/cutlass/cutlass_visitor_2x_broadcast_epilogue.hpp [hunk 9]
```diff
@@ -0,0 +1,332 @@
...
+      auto dst_v = filter(tC_rRow);
+      
+      if (params_ptr->ptr_row) {
+        CUTLASS_PRAGMA_UNROLL
+        for (int i = 0; i < size(src_v); ++i) {
+          bool guard = get<1>(coord_v(i)) < n;
+          cutlass::arch::global_load<VecType, sizeof(VecType)>(dst_v(i), (void const*)&src_v(i), guard);
+        }
+      } else {
+        VecType filled_vec;
```
Comment (LucasWilkinson) [inline csrc/quantization/cutlass/cutlass_visitor_2x_broadcast_epilogue.hpp:137]: maybe `broadcasted_vec` might be a better name them filled vec?

File: csrc/quantization/cutlass/cutlass_visitor_2x_broadcast_epilogue.hpp (hunk 4 above)
Comment (tlrmchlsmth) [reply csrc/quantization/cutlass/cutlass_visitor_2x_broadcast_epilogue.hpp]: Good call. This wasn't reading/writing to invalid memory but probably wasn't robust to reduction epilogues, so I've added some checks.

**Comment by tlrmchlsmth** [comment]
@jeejeelee I just added SM75 support as well. I didn't spent a ton of time tuning it but it's maybe 50% faster than fp16 GEMM

File: csrc/quantization/cutlass/scaled_mm_dq_c2x.cu [hunk 10]
```diff
@@ -0,0 +1,301 @@
...
+      ldc,
+      ldc};
+
+  // Launch the CUTLASS GEMM kernel.
+  typename Gemm::Op gemm_op;
+  size_t workspace_size = gemm_op.get_workspace_size(args);
+  cutlass::device_memory::allocation<uint8_t> workspace(workspace_size);
+
+  CUTLASS_CHECK(gemm_op.can_implement(args));
+  cutlass::Status status = gemm_op(args, workspace.get());
```
Comment (jeejeelee) [inline csrc/quantization/cutlass/scaled_mm_dq_c2x.cu:198]: Should we add  device guard here, like what was done in [device guard](https://github.com/vllm-project/vllm/blob/v0.4.2/csrc/layernorm_kernels.cu#L287-L288)

File: tests/kernels/test_cutlass.py [hunk 11]
```diff
@@ -0,0 +1,170 @@
...
+
+    scale_a = torch.randn((1, 1), device="cuda", dtype=torch.float32) / 10
+    scale_b = torch.randn((1, 1), device="cuda", dtype=torch.float32) / 10
+
+    out = ops.cutlass_scaled_mm_dq(a, b, scale_a, scale_b)
+    baseline = torch.mm(scale_a * a.to(dtype=torch.float32),
+                        scale_b *
+                        b.to(dtype=torch.float32)).to(dtype=torch.bfloat16)
+
+    assert torch.allclose(out, baseline, rtol=1e-1, atol=1e0)
```
Comment (jeejeelee) [inline tests/kernels/test_cutlass.py:170]: We should also consider whether  test can pass in non-zero GPU devices, similar to what is done in other kernel tests.

File: csrc/quantization/cutlass/scaled_mm_dq_c2x.cu [hunk 12]
```diff
@@ -0,0 +1,301 @@
+#include <assert.h>
```
Comment (jeejeelee) [inline csrc/quantization/cutlass/scaled_mm_dq_c2x.cu:1]: IMHO, perhaps quantization/cutlass_w8a8/ as folder name would be more appropriate. we may also need to be able to optimize other operators based on cutlass

File: tests/kernels/test_cutlass.py (hunk 11 above)
Comment (tlrmchlsmth) [reply tests/kernels/test_cutlass.py:196]: very good call, thanks!

File: csrc/quantization/cutlass_w8a8/scaled_mm_dq_c3x.cu [hunk 13]
```diff
@@ -0,0 +1,235 @@
...
+/////////////////////////////////////////
+
+/*
+   This defines a quantized GEMM operation with dequantized output, similar to
+   torch._scaled_mm. It is defined using the CUTLASS 3.x API, and is used for
+   NVIDIA GPUs with sm90a (Hopper) or later.
+
+   A and B may be either int8 or fp8_e4m3. A can be quantized per-tensor or
+   per-row. B can be quantized per-tensor or per-column. They must have
+   symmetric quantization.
```
Comment (pcmoritz) [inline csrc/quantization/cutlass_w8a8/scaled_mm_dq_c3x.cu:34]: It wasn't clear to me what "symmetric quantization" refers to here -- do you mean they either both need to be per-tensor or (per-row, per-column)?

Comment (tlrmchlsmth) [reply csrc/quantization/cutlass_w8a8/scaled_mm_dq_c3x.cu:34]: Sorry, I'll add some more detail to clarify this comment comment. Symmetric == zero point is 0.

We can mix and match per-row and per-column. The unit tests cover all four cases

File: csrc/quantization/cutlass_w8a8/scaled_mm_dq_c3x.cu [hunk 14]
```diff
@@ -0,0 +1,235 @@
...
+  using StrideC = StrideD;
+
+  using CollectiveEpilogue =
+      typename cutlass::epilogue::collective::CollectiveBuilder<
+          cutlass::arch::Sm90, cutlass::arch::OpClassTensorOp, TileShape,
+          ClusterShape, cutlass::epilogue::collective::EpilogueTileAuto,
+          ElementAcc, float, ElementC, StrideC, 4, ElementD, StrideD, 4,
+          EpilogueSchedule, EVTCompute1>::CollectiveOp;
+
+  static constexpr size_t CEStorageSize =
```
Comment (pcmoritz) [inline csrc/quantization/cutlass_w8a8/scaled_mm_dq_c3x.cu:100]: Nit: I found it to be slightly cleaner to inline this like in https://github.com/NVIDIA/cutlass/blob/main/examples/54_hopper_fp8_warp_specialized_gemm/54_hopper_fp8_warp_specialized_gemm.cu#L148 unless you have a reason to do that :)

File: csrc/quantization/cutlass_w8a8/scaled_mm_dq_c3x.cu (hunk 13 above)
Comment (pcmoritz) [reply csrc/quantization/cutlass_w8a8/scaled_mm_dq_c3x.cu:34]: Got it that makes sense -- I'm not as familiar with the int8 terminology unfortunately as I have mainly been working in fp8 :)

File: csrc/quantization/cutlass_w8a8/scaled_mm_dq_c3x.cu (hunk 14 above)
Comment (tlrmchlsmth) [reply csrc/quantization/cutlass_w8a8/scaled_mm_dq_c3x.cu:100]: LOL I spent hours on this actually! I initially had it the way that you suggest, but then I refactored the code to add the `cutlass_3x_gemm` class so that I could have common code for int8 and fp8 and template on the relevant hyperparams. 

When I did that, it failed to compile and emitted a ton of template error spam. Might have been an NVCC bug. It went away when I factored it out like this.

File: vllm/_custom_ops.py [hunk 15]
```diff
@@ -153,6 +153,36 @@ def marlin_gemm(a: torch.Tensor, b_q_weight: torch.Tensor,
...
 
 
+# cutlass
+def cutlass_scaled_mm_dq(
+        a: torch.Tensor,
+        b: torch.Tensor,
+        a_scales: torch.Tensor,
+        b_scales: torch.Tensor,
+        out_dtype: Type[torch.dtype] = torch.bfloat16) -> torch.Tensor:
+    shape_fallback = b.shape[0] % 16 != 0 or b.shape[1] % 16 != 0
```
Comment (pcmoritz) [inline vllm/_custom_ops.py:163]: I think it would be better to remove the fallback and just replace this with an assert (similar for `capability < 75`). The reason is that it can make performance unpredictable. For this low level method, the caller asks for `cutlass` (that's in the name), so they should get cutlass. Falling back should be done on a higher level if needed :)

File: vllm/_custom_ops.py [hunk 16]
```diff
@@ -153,6 +153,36 @@ def marlin_gemm(a: torch.Tensor, b_q_weight: torch.Tensor,
                                 size_n, size_k)
 
 
+# cutlass
+def cutlass_scaled_mm_dq(
+        a: torch.Tensor,
+        b: torch.Tensor,
+        a_scales: torch.Tensor,
+        b_scales: torch.Tensor,
+        out_dtype: Type[torch.dtype] = torch.bfloat16) -> torch.Tensor:
```
Comment (pcmoritz) [inline vllm/_custom_ops.py:162]: Shall we remove the default argument here? It is not clear that `torch.bfloat16` is more canonical than torch.float16 and can lead to bugs if the caller doesn't specify it -- better to be explicit here :)

File: csrc/quantization/cutlass_w8a8/scaled_mm_dq_c2x.cu [hunk 17]
```diff
@@ -0,0 +1,301 @@
...
+    assert(a_scales.dtype() == torch::kFloat32);
+    assert(b_scales.dtype() == torch::kFloat32);
+
+    if (out.dtype() == torch::kBFloat16) {
+      return cutlass_scaled_mm_dq_dispatcher<
+          cutlass_2x_gemm<cutlass::arch::Sm89, int8_t, cutlass::bfloat16_t,
+                          TileShape, WarpShape, InstructionShape, 5>>(
+          out, a, b, a_scales, b_scales);
+    } else {
+      return cutlass_scaled_mm_dq_dispatcher<
```
Comment (pcmoritz) [inline csrc/quantization/cutlass_w8a8/scaled_mm_dq_c2x.cu:277]: Add `assert(out.dtype() == torch::kFloat16);` above to be consistent with the rest of the code?

File: csrc/quantization/cutlass_w8a8/scaled_mm_dq_c3x.cu (hunk 14 above)
Comment (pcmoritz) [reply csrc/quantization/cutlass_w8a8/scaled_mm_dq_c3x.cu:100]: Oops, sorry to hear 😅

File: tests/kernels/test_cutlass.py [hunk 18]
```diff
@@ -0,0 +1,196 @@
...
+]
+
+capability = torch.cuda.get_device_capability()
+capability = capability[0] * 10 + capability[1]
+
+
+def to_fp8(tensor: torch.tensor):
+    # Assuming input tensor is float32
+    # Scale tensor to range of FP8 E4M3
+    # by clamping exponent and truncating mantissa
```
Comment (pcmoritz) [inline tests/kernels/test_cutlass.py:23]: Hehe I guess it works but the seems slightly odd -- why not go with a more traditional strategy like in https://github.com/vllm-project/vllm/blob/361c461a128a5df2faefeb70ffa98e61e4feda55/vllm/model_executor/layers/quantization/fp8.py#L261 and just clamp between min and max?

I guess the way it is right now will generate more diverse test data, but then in practice people will use the clamping between min and max (and that's also what you do for int8). You can use the same `* 5` that you use in cutlass_int8_gemm_helper in cutlass_fp8_gemm_helper too and that would probably be diverse enough without this odd conversion technique :)

File: vllm/_custom_ops.py (hunk 15 above)
Comment (tlrmchlsmth) [reply vllm/_custom_ops.py:163]: Agreed, much better this way

File: tests/kernels/test_cutlass.py (hunk 18 above)
Comment (tlrmchlsmth) [reply tests/kernels/test_cutlass.py:23]: yes much better :sweat_smile:

File: csrc/quantization/cutlass_w8a8/scaled_mm_dq_c2x.cu (hunk 17 above)
Comment (tlrmchlsmth) [reply csrc/quantization/cutlass_w8a8/scaled_mm_dq_c2x.cu:277]: Nice catch, thanks!

**Comment by pcmoritz** [comment]
Btw, while I was trying out this PR, I got the following error:

```python
import torch
from vllm import _custom_ops as ops

A = torch.randn(8, 4096, dtype=torch.float16, device="cuda")
B = torch.randn(4096, 8192, dtype=torch.float16, device="cuda")

A *= 500
B *= 500

def per_tensor_quantize(tensor: torch.Tensor,
                       inv_scale: float) -> torch.Tensor:
   finfo = torch.finfo(torch.float8_e4m3fn)
   qweight = (tensor / inv_scale).clamp(min=finfo.min, max=finfo.max)
   return qweight.to(torch.float8_e4m3fn)

A_scale = 448.0 / A.abs().max()
B_scale = 448.0 / B.abs().max()

Aquant = per_tensor_quantize(A, 1.0 / A_scale)
Bquant = per_tensor_quantize(B, 1.0 / B_scale)

scale_a = A_scale * torch.ones((1, 1), device="cuda")
scale_b = B_scale * torch.ones((1, 1), device="cuda")

out = ops.cutlass_scaled_mm_dq(Aquant, Bquant, scale_a, scale_b, out_dtype=torch.float16)

out
```

I'm getting
```
RuntimeError: CUDA error: misaligned address
CUDA kernel errors might be asynchronously reported at some other API call, so the stacktrace below might be incorrect.
For debugging consider passing CUDA_LAUNCH_BLOCKING=1.
Compile with `TORCH_USE_CUDA_DSA` to enable device-side assertions.
```

Can you have a look if you know what is happening here?

**Comment by pcmoritz** [comment]
Ah, I think I know -- I didn't transpose `B` appropriately :)

Can you add a check in `cutlass_scaled_mm_dq` to make sure the dimensions are compatible and the matrices are in the right format?

Otherwise the PR looks good to me! Happy to stamp after the above comments are addressed :)

**Comment by tlrmchlsmth** [comment]
I'll add some asserts :)

**Comment by pcmoritz** [comment]
Thanks! We should also assert that the tensors are contiguous :)

**Comment by tlrmchlsmth** [comment]
@pcmoritz I added some asserts. Please take another look. 

We actually don't want to assert that the tensors are contiguous. I specifically added a unit test for the case that we are working with submatrices of `A` and `B`. In that case `A` won't be contiguous either.  Also: I didn't know this but since we transpose `B`, `is_contiguous` will never be true.

**Comment by pcmoritz** [comment]
Did you push your changes? Most of my comments still need to be resolved :)

**Comment by tlrmchlsmth** [comment]
> Did you push your changes? Most of my comments still need to be resolved :)
should be there now :)

File: csrc/quantization/cutlass_w8a8/scaled_mm_dq_entry.cu [hunk 19]
```diff
@@ -0,0 +1,66 @@
...
+                          torch::Tensor const &b_scales) {
+  int32_t major_capability;
+  int32_t minor_capability;
+  cudaDeviceGetAttribute(&major_capability, cudaDevAttrComputeCapabilityMajor,
+                         0);
+  cudaDeviceGetAttribute(&minor_capability, cudaDevAttrComputeCapabilityMinor,
+                         0);
+  int32_t version_num = major_capability * 10 + minor_capability;
+
+  // Checks for conformality
```
Comment (pcmoritz) [inline csrc/quantization/cutlass_w8a8/scaled_mm_dq_entry.cu:37]: Can you make all of these TORCH_CHECK instead? Since we are not checking them in the Python code (most of them anyways), they could legitimately crash the interpreter which we should avoid :)

File: csrc/quantization/cutlass_w8a8/scaled_mm_dq_c2x.cu [hunk 20]
```diff
@@ -0,0 +1,298 @@
...
+  CUTLASS_CHECK(status);
+}
+
+}  // namespace
+
+void cutlass_scaled_mm_dq_sm75(torch::Tensor &out, torch::Tensor const &a,
+                               torch::Tensor const &b,
+                               torch::Tensor const &a_scales,
+                               torch::Tensor const &b_scales) {
+  assert(a.dtype() == torch::kInt8);
```
Comment (pcmoritz) [inline csrc/quantization/cutlass_w8a8/scaled_mm_dq_c2x.cu:204]: These asserts should be TORCH_CHECK as well :)

File: csrc/quantization/cutlass_w8a8/scaled_mm_dq_c2x.cu [hunk 21]
```diff
@@ -0,0 +1,298 @@
...
+                        WarpShape, InstructionShape, 2>>(out, a, b, a_scales,
+                                                         b_scales);
+  }
+}
+
+void cutlass_scaled_mm_dq_sm80(torch::Tensor &out, torch::Tensor const &a,
+                               torch::Tensor const &b,
+                               torch::Tensor const &a_scales,
+                               torch::Tensor const &b_scales) {
+  assert(a.dtype() == torch::kInt8);
```
Comment (pcmoritz) [inline csrc/quantization/cutlass_w8a8/scaled_mm_dq_c2x.cu:231]: ditto

File: csrc/quantization/cutlass_w8a8/scaled_mm_dq_c2x.cu [hunk 22]
```diff
@@ -0,0 +1,298 @@
...
+void cutlass_scaled_mm_dq_sm89(torch::Tensor &out, torch::Tensor const &a,
+                               torch::Tensor const &b,
+                               torch::Tensor const &a_scales,
+                               torch::Tensor const &b_scales) {
+  using TileShape = typename cutlass::gemm::GemmShape<128, 128, 64>;
+  using WarpShape = typename cutlass::gemm::GemmShape<64, 64, 64>;
+  using InstructionShape = typename cutlass::gemm::GemmShape<16, 8, 32>;
+
+  if (a.dtype() == torch::kInt8) {
+    assert(b.dtype() == torch::kInt8);
```
Comment (pcmoritz) [inline csrc/quantization/cutlass_w8a8/scaled_mm_dq_c2x.cu:263]: ditto

File: csrc/quantization/cutlass_w8a8/scaled_mm_dq_c2x.cu [hunk 23]
```diff
@@ -0,0 +1,298 @@
...
+          out, a, b, a_scales, b_scales);
+    } else {
+      assert(out.dtype() == torch::kFloat16);
+      return cutlass_scaled_mm_dq_dispatcher<
+          cutlass_2x_gemm<cutlass::arch::Sm89, int8_t, cutlass::half_t,
+                          TileShape, WarpShape, InstructionShape, 5>>(
+          out, a, b, a_scales, b_scales);
+    }
+  } else {
+    assert(a.dtype() == torch::kFloat8_e4m3fn);
```
Comment (pcmoritz) [inline csrc/quantization/cutlass_w8a8/scaled_mm_dq_c2x.cu:280]: here as well

**Comment by pcmoritz** [comment]
Thanks for the fixes, I have a few more comments!

As a mental picture, it should never be possible to crash the python interpreter from python code. Asserts in the C++ level should only be used for consistency checks with previously already established invariants, never for input validations :)

**Comment by pcmoritz** [comment]
Otherwise the PR looks good to me now :)

**Comment by tlrmchlsmth** [comment]
Should be ready now, thanks! @pcmoritz

**Review by pcmoritz — APPROVED** [review]

**Comment by brisker** [comment]
@tlrmchlsmth 
Any plan on w4a8 quantization support?

**Comment by brisker** [comment]
@tlrmchlsmth 
Any plan on w4a8 quantization support?

**Comment by brisker** [comment]
@tlrmchlsmth 
Any plan on w4a8 quantization support?

**Comment by shesung** [comment]
Is there some benchmark results for w8a8 speedup?

**Comment by MuYu-zhi** [comment]
@tlrmchlsmth hi, I'm invoking the `cutlass_scaled_mm_dq` kernel with `enforce_eager=False`mode, and raising an error, 

> [rank0]:   File "/gjr/vllm/vllm/_custom_ops.py", line 189, in cutlass_scaled_mm_dq
[rank0]:     vllm_ops.cutlass_scaled_mm_dq(out, a, b, a_scales, b_scales)
[rank0]: RuntimeError: CUDA error: operation not permitted when stream is capturing
[rank0]: Compile with `TORCH_USE_CUDA_DSA` to enable device-side assertions.

It seems the implementation of this kernel is not compatible with cuda graph.

Do you have any advice for this? 3q

**Comment by tlrmchlsmth** [comment]
Hey @shesung, if you are looking for end-to-end results for w8a8, we do use the CUTLASS kernels for the fp8 results here https://twitter.com/neuralmagic/status/1812863986330910816

**Comment by tlrmchlsmth** [comment]
@MuYu-zhi are you using the kernels from this PR directly? In that case, yes they did not initially support CUDA graphs. They were also completely untuned and slow in their initial version, so I'd recommend looking at the ones from vLLM main

**Comment by MuYu-zhi** [comment]
@tlrmchlsmth I  pulled from vllm main, but not the latest main, it's version 0.4.2. Does the kernel have any updates after 0.4.2? If I want to support cuda graph by myself, how?  I don't have extensive experience in cuda graph.

**Comment by robertgshaw2-neuralmagic** [comment]
@MuYu-zhi yes, you need to upgrade to a newer version of vllm

Is there a reason you need to use 0.4.2?

**Comment by MuYu-zhi** [comment]
@robertgshaw2-neuralmagic No specific reason, it's just that the latest version was 0.4.2 when I pulled it at that time. I'll try updating it. Thanks.
